drivers/hmis/app/graphql/mutations/ce/submit_ce_referral_step.rb

```diff
@@ -16,23 +16,64 @@ class Ce::SubmitCeReferralStep < CleanBaseMutation
     argument :step_id, ID, required: true
     argument :input, Types::JsonObject, required: true
-    field :step, Types::HmisSchema::CeReferralStep, null: false
-    field :referral, Types::HmisSchema::CeReferral, null: false # return the referral so the UI can respond appropriately if the referral's overall status has changed
+    argument :confirmed, Boolean, required: false
 
-    def resolve(referral_id:, step_id:, input:)
+    field :step, Types::HmisSchema::CeReferralStep, null: true # nullable in case of errors
+    field :referral, Types::HmisSchema::CeReferral, null: true # return the referral so the UI can respond appropriately if the referral's overall status has changed
+
+    def resolve(referral_id:, step_id:, input:, confirmed: false)
       raise unless Hmis::Ce.configuration.enabled?
 
       referral = Hmis::Ce::Referral.viewable_by(current_user).find(referral_id)
-      step = nil
+      errors = HmisErrors::Errors.new
+
+      unless confirmed
+        # dependency injection style engine
+        dry_run_engine = Hmis::WorkflowExecution::Engine.new(
+          referral.workflow_instance,
+          message_handler: Hmis::Ce::DryRunMessageHandler.new,
+          assignment_handler: nil, # dry runner does not do any assignment
+          stepper: Hmis::Ce::DryRunEngineStepper.new,
+          audit_logger: nil, # or log audit events
+        )
+
+        step = dry_run_engine.active_steps.find(step_id)
+
+        errors.push(*validate(step, input))
+        return { errors: errors } if errors.any? # return early - if there are validation errors, they may cause the dry run to fail
+
+        dry_run_engine.complete_step!(step, user: current_user, submitted_values: input)
+
+        # todo @martha - this isn't flexible enough, it needs to be configurable which types of messages require a warning.
+        # example with admin denials - we don't need to warn the admin that they are moving the referral to 'rejected,' we need to warn the non-admin that they are sending the referral to an admin for review
+
+        if dry_run_engine.message_handler.collected_messages.map(&:type).include?('reject_referral')
+          errors.add(:base, :information, full_message: 'This will decline the referral', severity: :warning)
+          return { errors: errors }
+        end
+      end
+
       referral.opportunity.with_lock do
         engine = referral.workflow_engine
         step = engine.active_steps.find(step_id)
+
+        errors.push(*validate(step, input))
+        return { errors: errors } if errors.any?
+
         engine.complete_step!(step, user: current_user, submitted_values: input)
       end
+
       {
         step: step,
         referral: referral.reload,
       }
     end
+
+    def validate(step, submitted_values)
+      definition = step.node.form_definition
+      return unless definition && submitted_values
+
+      definition.validate_form_values(submitted_values)
+    end
```

Comment:
as discussed at dev standup - makes sense to move this logic into the step

Reply:
Thanks yep! I've added a new method on the engine, rather than doing this as part of `complete_step`, in order to more gracefully respond to it in the mutation. This was roughly based on the submit form mutation, which [collects validation from the form processor](https://github.com/greenriver/hmis-warehouse/blob/ab8d55969a9a737fe4df3bbd6f44d593a5e9d116/drivers/hmis/app/graphql/mutations/submit_form.rb#L68-L71) before running. Let me know if you think it looks okay.